### PR TITLE
Cherry pick changes up to Feb 18

### PR DESCRIPTION
**Pull request by jarcherNV**
Author: William Lew <wlew@nvidia.com>

ARM64 Support

Adds full ARM64 support for running AlpaSim on Grace-Blackwell platforms.

Tested on the following hardware:
- **DGX Spark** - local Docker Compose
- **DGX Station** - local Docker Compose
- **SLURM cluster** - pyxis/enroot

All services now run on ARM: driver, physics, controller, runtime, and sensorsim (NRE with gsplat).
Only trafficsim is skipped.

| Workaround | Reason |
|---|---|
| NGC PyTorch base image | Only source of CUDA-enabled PyTorch on aarch64 (no PyPI wheels) |
| VaVAM FP32 + no autocast | `torch.amp.autocast` doesn't auto-cast Float32 weights in F.linear on ARM — confirmed broken on PyTorch 2.6 and 2.8 |
| `TORCH_CUDA_ARCH_LIST=9.0+PTX` on sensorsim | Blackwell compute 10.3 not recognized by NRE's PyTorch; slangtorch JIT fails |
| `--container-writable` on srun | IPP5 pyxis defaults to read-only; uv/matplotlib need writable filesystem |
| gsplat renderer (not nrend) | NRE nrend C++ renderer expects newer scene data format |
| Pin setuptools<82 | grpcio-tools imports pkg_resources, removed in setuptools 82 |

---
Author: Jesse Archer <jarcher@nvidia.com>

cicd: make test-slurm wait for build-services before on-demand squash
When the wizard creates .sqsh on demand it runs enroot import; the image
must exist in the registry. Add build-services as an optional need for
test-slurm so the image is built and pushed before the SLURM test runs.

Trivial comment in src/runtime/pyproject.toml to trigger build-services
in CI for verifying this fix.

---
Author: Jesse Archer <jarcher@nvidia.com>

wizard: create squash files on demand
Remove the separate squash-services and squash-external-services CI
jobs. The wizard now ensures .sqsh files exist when deploying on SLURM:
it looks in wizard.sqshcaches and, if missing, creates the squash in
the first writable cache under a per-image file lock. Per-cluster
cache remains configurable via defines.filesystem / sqshcaches.

- Add ensure_sqsh_path() in wizard utils (lock + enroot import).
- Use ensure_sqsh_path in slurm deployment instead of image_url_to_sqsh_filename.
- Remove squash job template and squash-services / squash-external-services.
- test-slurm no longer depends on squash jobs; drop SQUASH_CACHE_DIR check.
- Dev image tags: optional lockfile hash (uv.lock) for version isolation.
- Remove .manual_pipeline_squash and "Squash Images" pipeline option.

---
Author: Michal Tyszkiewicz <mtyszkiewicz@nvidia.com>

[physics] remove point_cloud_utils dependency
`point_cloud_utils` is a behemoth that depends transitively on x86-only intrinsics making it impossible to compile for arm64. Since we only use its `.ply` I/O capabilities, I replace it with pure numpy I/O.

Alternatives:

* Use \[plyfile\](https://pypi.org/project/plyfile/) - gpl only :disappointed:
* Use various less popular libraries like \[pyminiply\](https://pypi.org/project/pyminiply/) - may not support writing and require us to build the linux_aarch64 wheels ourselves

This impl may be less performance than native alternatives but on the other hand it reads directly from `bytes`, skipping the need for a `NamedTemporaryFile` which most other libraries require.

---
Author: Maximilian Igl <migl@nvidia.com>

chore: add .worktrees to gitignore
For trying out various Claude code setups.

---
Author: Maximilian Igl <migl@nvidia.com>

driver: replace scipy interpolation with Rust-backed Trajectory
- Add `rotation_matrices` getter to `utils_rs.Trajectory` — returns `(N, 3, 3)` numpy array of rotation matrices via `glam::Mat3::from_quat`
- Refactor driver's `_build_ego_history` to use `Trajectory.interpolate()` + `.transform()` + `.rotation_matrices` instead of scipy's `interp1d`/`Slerp` and the manual `_adjust_orientation` hemisphere fix
- Extract `build_ego_history` as a standalone, testable function
- Add 9 tests covering straight-line, turning, and antipodal quaternion scenarios

glam's `slerp` handles quaternion antipodal cases natively (checks dot product per-pair), eliminating the need for the manual `_adjust_orientation` workaround required by scipy's global `Slerp`. Net result: ~100 lines of scipy boilerplate replaced with ~30 lines of Trajectory API calls.
Additionally, the Rust implementation should be significantly faster.

---
Author: Maximilian Igl <migl@nvidia.com>

Batching of driver
Batch Vavam and Transfuser Inference.

---
Author: Maximilian Igl <migl@nvidia.com>

Check in Claude.md
We have access to Claude code, which doesn't natively read AGENTS.md

This auto-loads AGENTS.md from CLAUDE.md

---
Author: Maximilian Igl <migl@nvidia.com>

utils_rs: validate and normalize unit quaternions at Rust boundary
The Rust `Pose` and `Trajectory` types use `glam::Quat` internally, which assumes unit quaternions for all operations (slerp, inverse, compose, rotation matrix conversion). However, none of the entry points validated or normalized incoming quaternions, meaning non-unit quaternions would silently produce wrong results.

Add a validate-and-normalize check at the Rust/Python boundary for all quaternion entry points:

- **Within tolerance** (length² deviation ≤ 1e-2): silently normalized to handle float drift from f64→f32 conversion, serialization round-trips, etc.
- **Near-zero**: rejected with clear error (corrupt/uninitialized data)
- **Far from unit**: rejected with clear error (likely caller bug)

Entry points covered: `Pose::new`, `Pose::from_proto`, `Trajectory::new`.

Also:
- Removed now-redundant Python-side validation in `camera_catalog.py`
- Fixed `pose_from_grpc` to handle the protobuf default-value edge case (unset `Quat` → `[0,0,0,0]` → identity)
- Added 9 tests for the new validation behavior

---
Author: Maximilian Igl <migl@nvidia.com>

driver: Simplify predict interface and enable batching
- Bundle the 5 positional parameters of `BaseTrajectoryModel.predict()` into a single `PredictionInput` dataclass. Introduce `CameraFrame` (NamedTuple) and `CameraImages` type alias to replace the opaque `dict[str, list[tuple[int, np.ndarray]]]`.
- Add `predict_batch()` to `BaseTrajectoryModel` with a default sequential implementation. Models can override this for GPU-level batching. The servicer's `_run_batch` now delegates to `predict_batch`.
- Fix VaVAM test asset path (`data/drivers/` → `data/drivers/vavam/`).

The previous `predict` signature had kitchen-sink parameters (`speed`, `acceleration`, `ego_pose_history`) where each model used a different subset and `del`-ed the rest. The nested `dict[str, list[tuple[int, np.ndarray]]]` type was hard to read. There was no extension point for GPU batching despite existing queue infrastructure.

- [x] All 13 driver tests pass (including VAM end-to-end service flow on GPU)
- [ ] CI pipeline

---
Author: Peter Karkus <pkarkus@nvidia.com>

Introduce video layouts, and add Alpamayo 1 reasoning video overlay
Changes are pulled in from public branch

---
Author: Michal Tyszkiewicz <mtyszkiewicz@nvidia.com>

08c10bd3 Reimplement perf-critical parts of `alpasim_utils` in Rust
Reimplements `Trajectory` and `Polyline` as native objects to improve performance - profiling has shown that `Trajectory.interpolate_to_timestamps` and `Polyline.project_point` are substantial sources of CPU overhead. To avoid an awkward python/rust split, I port most other methods to rust as well.

* A new "scalar" type called `Pose` (xyz + quat) is implemented and `QVec` (understood as a batch of `Pose` objects) is removed entirely. The vast majority of uses of `QVec` was to construct a `Trajectory` immediately.
* Rust-native types cannot be subclassed in Python. At the same time, some methods which use libs other than `numpy` are involved to port to rust. This is in particular `grpc_types` conversion and some spline handling code. For this reason, I made such methods "free" functions, formally separate from the native type. This means e.g. `trajectory_to_grpc(grpc_types.Trajectory)` instead of `Trajectory.from_grpc(grpc_types.Trajectory)`.
  *  There may be some options to enable subclassing that I haven't looked into yet.
* Installation/development flow (see below)

* The rust bindings are created with the `pyo3` library and the whole thing is exposed as a package via the `maturin` build backend.
* `maturin` requires that the user has rust compiler ([installation](https://rust-lang.org/tools/install/)) on their system **to build the extension**.
  * For the average user, "their system" means the `alpasim-base` docker image so the change is internal. I have added that dependency there.
* TODO: development flow with auto-rebuilds

---
Author: Maximilian Igl <migl@nvidia.com>

docs: Update CODEBASE_MAP and AGENTS.md
Regenerates `docs/CODEBASE_MAP.md` with comprehensive module documentation, data flow diagrams, and navigation guides. Updates `AGENTS.md` with key architectural patterns.

**Changes:**
- Enhanced system overview with Mermaid diagrams
- Detailed module guide (Runtime, Wizard, Driver, Controller, Physics, Eval, gRPC, Utils, Tools)
- Data flow sequences, worker parallelism docs, coordinate frames table
- Conventions, gotchas, and navigation guide
- Reference to `RUNTIME_PARALLELISM.md` for parallelism details

---
Author: Maximilian Igl <migl@nvidia.com>

Add doc-reviewer skill
Mostly AI generated (with feedback & iteration) skill.
Idea is that for each signficiant change in the codebase, you call this skill from cursor to check if the documentation needs updating.

If you find that the skill does not work as intended, please iterate on the SKILL.md itself!

### CODE REVIEW

File: docs/agents/CODEBASE_MAP.md
Comment (mwatson-nvidia) [inline docs/agents/CODEBASE_MAP.md:1]: maybe drop all the docs/agents files since I think we're dropping these soon?

File: src/grpc/tutorial.md [hunk 1]
```diff
@@ -37,4 +37,92 @@ Number 1. is for **clients**, 2. and 3. are for the **server**.
 Unfortunately, the bare gRPC codegen doesn't produce valid Python packages - without `__init__.py` and relative imports you can't just output the generated files to an arbitrary location in your codebase and import them from there. This is the reason for this package, which has a well defined root, contains "hand-made" `__init__.py` and after installing allows imports like `from alpasim_grpc.v0.your_service_pb2_grpc import your_thing`.
 
 ## Implementing
+<<<<<<< HEAD
```
Comment (mwatson-nvidia) [inline src/grpc/tutorial.md]: unresolved conflict? I'd probably remove most of this stuff

File: src/wizard/alpasim_wizard/utils.py [hunk 2]
```diff
@@ -68,6 +78,108 @@ def image_url_to_sqsh_filename(image: str, squash_caches: list[str]) -> str:
     raise ValueError(f"Could not find file: {sqsh_fname} at {sqsh_paths=}.")
 
 
+def _image_to_enroot_uri(image: str) -> str:
+    """Convert docker image URL to enroot URI with auth placeholder (nvcr.io)."""
+    # enroot reads $oauthtoken from credentials; pass literal so enroot can substitute
+    if image.startswith("nvcr.io/"):
+        return "docker://$oauthtoken@nvcr.io#" + image[len("nvcr.io/") :]
```
Comment (mwatson-nvidia) [inline src/wizard/alpasim_wizard/utils.py]: the nvcr.io stuff doesn't seem like it should be on the public branch

File: AGENTS.md [hunk 3]
```diff
@@ -42,7 +80,36 @@
 - Extend async tests with `pytest-asyncio`; keep gRPC client stubs isolated to avoid network
   side-effects.
 
+## Architecture & Data Flow
+- Services: **Runtime**, **Physics**, **Controller**, **Eval**, **DDB**, **Wizard**, **GRPC**, **Utils** (see `src/<service>/`).
+- All services communicate via gRPC messages defined in `src/grpc/alpasim_grpc/v0/`.
+- Wizard generates configs, Runtime orchestrates loop, services exchange data, Runtime logs to ASL, Eval computes metrics, DDB indexes outputs.
+
+## Environment Variables
+- Required: `ALPAMAYO_S3_SECRET`.
```
Comment (mwatson-nvidia) [inline AGENTS.md]: this, at least, seems internal only

File: src/grpc/tutorial.md (hunk 1 above)
Comment (jarcherNV) [reply src/grpc/tutorial.md]: Whoops, somehow I missed this one. It should get covered by the upcoming docs change so it's probably find to just remove it.

File: src/wizard/alpasim_wizard/utils.py (hunk 2 above)
Comment (jarcherNV) [reply src/wizard/alpasim_wizard/utils.py]: Good point. I will remove it.

File: AGENTS.md (hunk 3 above)
Comment (jarcherNV) [reply AGENTS.md]: I will remove this too.

File: docs/agents/CODEBASE_MAP.md
Comment (jarcherNV) [reply docs/agents/CODEBASE_MAP.md:1]: Done.
